### PR TITLE
Bug/#204 user is able to click 'signup' multiple times

### PR DESCRIPTION
**Pull request by EspressoCode**
# Related Issue
- #204 user is able to click 'signup' multiple times

# Proposed changes
This pr fixes the problem of multiple clicks on signup and login buttons. Now when the user clicks on sign up or login buttons, the button become unclickable (disabled) and it becomes available only in case that we have an appError which means the signup/login forms are filled out with valid entries but the response from backend shows that authentication/user creation failed.

# Reviewer(s)
 - @Alessandro100

### CODE REVIEW

File: client/src/app/components/signup-form/signup-form.component.html [hunk 1]
```diff
@@ -22,14 +22,17 @@ <h1 class="signup-title">Sign Up</h1>
...
 
-        <button id="signup-submit-btn" class="mt-input-field" mat-flat-button type="submit" [class.btn-opacity-disabled]="signUpForm.invalid"
-            [disabled]="!signUpForm.valid">
+        <button id="signup-submit-btn" class="mt-input-field" mat-flat-button type="submit"
+            [class.btn-opacity-disabled]="signUpForm.invalid"
+            [disabled]="(!signUpForm.valid || submitted) && !appError">
             SIGN UP
         </button>
-
+        <div class="form-error-messages" *ngIf="appError">A user with this email address or username already exist!
```
Comment (arthur-del) [inline client/src/app/components/signup-form/signup-form.component.html:34]: Should we specify which between the username and the email  already exists? I feel as though signing up should be as smooth as possible of process.

File: client/src/app/components/login-form/login-form.component.html [hunk 2]
```diff
@@ -8,8 +8,9 @@ <h2 class="login-text-hint">Already have an account?</h2>
     <mat-form-field class="mt-input-field half" appearance="outline" color="accent">
       <input id="login-password" matInput type="password" placeholder="Password" [formControl]="pwd" />
     </mat-form-field>
-    <button id="login-submit-btn" mat-flat-button class="mt-input-field half" [class.btn-opacity-disabled]="logInForm.invalid"
-      [disabled]="logInForm.invalid">LOGIN</button>
+    <button id="login-submit-btn" mat-flat-button class="mt-input-field half"
+      [class.btn-opacity-disabled]="logInForm.invalid"
+      [disabled]="(logInForm.invalid || submitted) && !appError">LOGIN</button>
```
Comment (Alessandro100) [inline client/src/app/components/login-form/login-form.component.html:13]: on line 13, there is a lot of logic going on in the [disabled] box, we can move this to a function and test it.

File: client/src/app/components/signup-form/signup-form.component.html [hunk 3]
```diff
@@ -22,14 +22,17 @@ <h1 class="signup-title">Sign Up</h1>
...
-            <input id="signup-password-2" matInput placeholder="Confirm Password" type="password" [formControl]="pwd2" />
+            <input id="signup-password-2" matInput placeholder="Confirm Password" type="password"
+                [formControl]="pwd2" />
         </mat-form-field>
 
-        <button id="signup-submit-btn" class="mt-input-field" mat-flat-button type="submit" [class.btn-opacity-disabled]="signUpForm.invalid"
-            [disabled]="!signUpForm.valid">
+        <button id="signup-submit-btn" class="mt-input-field" mat-flat-button type="submit"
+            [class.btn-opacity-disabled]="signUpForm.invalid"
+            [disabled]="(!signUpForm.valid || submitted) && !appError">
```
Comment (Alessandro100) [inline client/src/app/components/signup-form/signup-form.component.html:31]: Same comment as above, there is a lot of logic going on here that we can test

**Review by Alessandro100 — CHANGES_REQUESTED** [review]
I left some changes regarding refactoring / testing logic. Also I noticed that some of the errors are overlapping, we should look into that
image

**Comment by Alessandro100** [comment]
oh another thing, don't use quotation marks in a branch name! It wont be simple to copy past since we will have to do \'

**Comment by EspressoCode** [comment]
> oh another thing, don't use quotation marks in a branch name! It wont be simple to copy past since we will have to do\ '


ok sorry, I will keep that in mind for next time.

File: client/src/app/components/signup-form/signup-form.component.html (hunk 1 above)
Comment (EspressoCode) [reply client/src/app/components/signup-form/signup-form.component.html]: that is the right way, I agree! But for now, we have a generic message for both, in future, we will separate the two after getting this one working properly.

**Comment by EspressoCode** [comment]
I used the new code in my PR for error handling for fixing the issue of multiple clicks, please review that PR first, so that if there are changes to that one I can apply them here. I resolved all the issues and moved a part of the logic to functions, I will add tests for new parts soon and test login/sign up thoroughly again this time.

**Review by Alessandro100 — APPROVED** [review]
Looks good, nice changes

**Pull request merged**